File: frontend/mobile/google/android/MyHomeAgent/app/src/main/java/com/hestami_ai/myhomeagent/data/model/Property.kt
package com.hestami_ai.myhomeagent.data.model

import com.google.gson.annotations.SerializedName
import java.util.Date

/**
 * Property status enumeration matching iOS PropertyStatus.
 */
enum class PropertyStatus {
    @SerializedName("ACTIVE") ACTIVE,
    @SerializedName("INACTIVE") INACTIVE,
    @SerializedName("PENDING") PENDING,
    @SerializedName("COUNTY_PROCESSING") COUNTY_PROCESSING,
    UNKNOWN;

    companion object {
        @Suppress("unused")
        fun fromString(value: String): PropertyStatus {
            return entries.find { it.name == value } ?: UNKNOWN
        }
    }
}

/**
 * Property utilities information.
 */
data class PropertyUtilities(
    @SerializedName("gas")
    val gas: String? = null,

    @SerializedName("sewer")
    val sewer: String? = null,

    @SerializedName("water")
    val water: String? = null,

    @SerializedName("electricity")
    val electricity: String? = null,

    @SerializedName("internetCable")
    val internetCable: String? = null
)

/**
 * Property water heater information.
 */
data class PropertyWaterHeater(
    @SerializedName("type")
    val type: String? = null,

    @SerializedName("fuel")
    val fuel: String? = null,

    @SerializedName("capacity")
    val capacity: String? = null,

    @SerializedName("age")
    val age: Int? = null,

    @SerializedName("brand")
    val brand: String? = null,

    @SerializedName("model")
    val model: String? = null
)

/**
 * Property appliances information.
 */
data class PropertyAppliances(
    @SerializedName("refrigerator")
    val refrigerator: Boolean? = null,

    @SerializedName("stove")
    val stove: Boolean? = null,

    @SerializedName("dishwasher")
    val dishwasher: Boolean? = null,

    @SerializedName("washer")
    val washer: Boolean? = null,

    @SerializedName("dryer")
    val dryer: Boolean? = null,

    @SerializedName("microwave")
    val microwave: Boolean? = null,

    @SerializedName("oven")
    val oven: Boolean? = null,

    @SerializedName("range")
    val range: Boolean? = null,

    @SerializedName("garbageDisposal")
    val garbageDisposal: Boolean? = null
)

/**
 * Geocode position coordinates.
 */
data class GeocodePosition(
    @SerializedName("lat")
    val lat: Double? = null,

    @SerializedName("lon")
    val lon: Double? = null
)

/**
 * Geocode result address.
 */
data class GeocodeResultAddress(
    @SerializedName("country")
    val country: String? = null,

    @SerializedName("localName")
    val localName: String? = null,

    @SerializedName("postalCode")
    val postalCode: String? = null,

    @SerializedName("streetName")
    val streetName: String? = null,

    @SerializedName("countryCode")
    val countryCode: String? = null,

    @SerializedName("municipality")
    val municipality: String? = null,

    @SerializedName("streetNumber")
    val streetNumber: String? = null,

    @SerializedName("freeformAddress")
    val freeformAddress: String? = null
)

/**
 * Geocode result.
 */
data class GeocodeResult(
    @SerializedName("id")
    val id: String? = null,

    @SerializedName("type")
    val type: String? = null,

    @SerializedName("score")
    val score: Double? = null,

    @SerializedName("address")
    val address: GeocodeResultAddress? = null,

    @SerializedName("position")
    val position: GeocodePosition? = null
)

/**
 * Geocode summary.
 */
data class GeocodeSummary(
    @SerializedName("query")
    val query: String? = null,

    @SerializedName("queryTime")
    val queryTime: Int? = null
)

/**
 * Geocode address containing results and summary.
 */
data class GeocodeAddress(
    @SerializedName("results")
    val results: List<GeocodeResult>? = null,

    @SerializedName("summary")
    val summary: GeocodeSummary? = null
)

/**
 * Property descriptives containing detailed property information.
 * Matches iOS PropertyDescriptives implementation.
 */
data class PropertyDescriptives(
    // Basic Property Info
    @SerializedName("propertyType")
    val propertyType: String? = null,

    @SerializedName("yearBuilt")
    val yearBuilt: Int? = null,

    @SerializedName("squareFootage")
    val squareFootage: Int? = null,

    @SerializedName("lotSize")
    val lotSize: String? = null,

    @SerializedName("stories")
    val stories: Int? = null,

    @SerializedName("bedrooms")
    val bedrooms: Int? = null,

    @SerializedName("bathrooms")
    val bathrooms: String? = null,

    @SerializedName("unitNumber")
    val unitNumber: String? = null,

    // Access & Security
    @SerializedName("gatedCommunity")
    val gatedCommunity: Boolean? = null,

    @SerializedName("accessCode")
    val accessCode: String? = null,

    @SerializedName("accessInstructions")
    val accessInstructions: String? = null,

    @SerializedName("parkingType")
    val parkingType: String? = null,

    @SerializedName("parkingSpaces")
    val parkingSpaces: Int? = null,

    // Structure & Features
    @SerializedName("basement")
    val basement: Boolean? = null,

    @SerializedName("basementType")
    val basementType: String? = null,

    @SerializedName("garage")
    val garage: Boolean? = null,

    @SerializedName("garageType")
    val garageType: String? = null,

    @SerializedName("garageSpaces")
    val garageSpaces: Int? = null,

    @SerializedName("attic")
    val attic: Boolean? = null,

    @SerializedName("atticAccess")
    val atticAccess: String? = null,

    @SerializedName("crawlSpace")
    val crawlSpace: Boolean? = null,

    // HVAC & Climate Control
    @SerializedName("heatingSystem")
    val heatingSystem: String? = null,

    @SerializedName("heatingFuel")
    val heatingFuel: String? = null,

    @SerializedName("coolingSystem")
    val coolingSystem: String? = null,

    @SerializedName("airConditioning")
    val airConditioning: Boolean? = null,

    @SerializedName("hvacAge")
    val hvacAge: Int? = null,

    @SerializedName("hvacBrand")
    val hvacBrand: String? = null,

    @SerializedName("hvacModel")
    val hvacModel: String? = null,

    @SerializedName("thermostatType")
    val thermostatType: String? = null,

    // Utilities & Systems
    @SerializedName("utilities")
    val utilities: PropertyUtilities? = null,

    @SerializedName("waterSource")
    val waterSource: String? = null,

    @SerializedName("sewerSystem")
    val sewerSystem: String? = null,

    @SerializedName("electricalPanel")
    val electricalPanel: String? = null,

    @SerializedName("electricalAmps")
    val electricalAmps: Int? = null,

    @SerializedName("gasService")
    val gasService: Boolean? = null,

    @SerializedName("waterHeater")
    val waterHeater: PropertyWaterHeater? = null,

    // Plumbing
    @SerializedName("plumbingType")
    val plumbingType: String? = null,

    @SerializedName("waterShutoffLocation")
    val waterShutoffLocation: String? = null,

    @SerializedName("mainDrainCleanout")
    val mainDrainCleanout: String? = null,

    // Electrical
    @SerializedName("electricalPanelLocation")
    val electricalPanelLocation: String? = null,

    @SerializedName("wiringType")
    val wiringType: String? = null,

    // Roofing & Exterior
    @SerializedName("roofType")
    val roofType: String? = null,

    @SerializedName("roofAge")
    val roofAge: String? = null,

    @SerializedName("exteriorMaterial")
    val exteriorMaterial: String? = null,

    @SerializedName("foundationType")
    val foundationType: String? = null,

    // Appliances
    @SerializedName("appliances")
    val appliances: PropertyAppliances? = null,

    // Smart Home & Internet
    @SerializedName("internetService")
    val internetService: Boolean? = null,

    @SerializedName("internetProvider")
    val internetProvider: String? = null,

    @SerializedName("smartHomeDevices")
    val smartHomeDevices: List<String>? = null,

    // Landscaping & Exterior
    @SerializedName("sprinklerSystem")
    val sprinklerSystem: Boolean? = null,

    @SerializedName("pool")
    val pool: Boolean? = null,

    @SerializedName("poolType")
    val poolType: String? = null,

    @SerializedName("fence")
    val fence: Boolean? = null,

    @SerializedName("fenceType")
    val fenceType: String? = null,

    @SerializedName("deck")
    val deck: Boolean? = null,

    @SerializedName("deckMaterial")
    val deckMaterial: String? = null,

    @SerializedName("patio")
    val patio: Boolean? = null,

    @SerializedName("patioMaterial")
    val patioMaterial: String? = null,

    // Special Considerations
    @SerializedName("petFriendly")
    val petFriendly: Boolean? = null,

    @SerializedName("smokingAllowed")
    val smokingAllowed: Boolean? = null,

    @SerializedName("wheelchairAccessible")
    val wheelchairAccessible: Boolean? = null,

    @SerializedName("fireplace")
    val fireplace: Boolean? = null,

    @SerializedName("fireplaceType")
    val fireplaceType: String? = null,

    // Maintenance & Notes
    @SerializedName("maintenanceNotes")
    val maintenanceNotes: String? = null,

    @SerializedName("specialInstructions")
    val specialInstructions: String? = null,

    // Legacy field
    @SerializedName("created_from")
    val createdFrom: String? = null
)

/**
 * Property model matching iOS Property.swift implementation.
 */
data class Property(
    @SerializedName("id")
    val id: String,

    @SerializedName("title")
    val title: String,

    @SerializedName("description")
    val description: String,

    @SerializedName("address")
    val address: String,

    @SerializedName("city")
    val city: String,

    @SerializedName("state")
    val state: String,

    @SerializedName("zip_code")
    val zipCode: String,

    @SerializedName("county")
    val county: String? = null,

    @SerializedName("country")
    val country: String,

    @SerializedName("status")
    val status: PropertyStatus,

    @SerializedName("created_at")
    val createdAt: Date,

    @SerializedName("updated_at")
    val updatedAt: Date,

    @SerializedName("owner")
    val owner: String,

    @SerializedName("owner_details")
    val ownerDetails: User,

    @SerializedName("media_count")
    val mediaCount: Int,

    @SerializedName("descriptives")
    val descriptives: PropertyDescriptives? = null,

    @SerializedName("service_requests")
    val serviceRequests: List<ServiceRequest>? = null,

    @SerializedName("geocode_address")
    val geocodeAddress: GeocodeAddress? = null,

    @SerializedName("geocode_address_source")
    val geocodeAddressSource: String? = null,

    @SerializedName("permit_retrieval_status")
    val permitRetrievalStatus: String? = null,

    @SerializedName("permit_last_retrieved_at")
    val permitLastRetrievedAt: Date? = null,

    @SerializedName("permit_retrieval_error")
    val permitRetrievalError: String? = null,

    @SerializedName("permit_next_retrieval_at")
    val permitNextRetrievalAt: Date? = null,

    @SerializedName("permit_retrieval_workflow_id")
    val permitRetrievalWorkflowId: String? = null,

    @SerializedName("featuredImage")
    val featuredImage: String? = null,

    @SerializedName("media")
    val media: List<Media>? = null
) {
    /**
     * Full address string for display.
     */
    @Suppress("unused")
    val fullAddress: String
        get() = "$address, $city, $state $zipCode"
}

File: frontend/mobile/apple/iOS/My Home Agent/New Group/My Home Agent/Models/Property.swift
import Foundation
import UIKit

// Helper extension for flexible Int decoding (handles both Int and String from backend)
extension KeyedDecodingContainer {
    func decodeFlexibleInt(forKey key: K) throws -> Int? {
        // Try to decode as Int first
        if let intValue = try? decodeIfPresent(Int.self, forKey: key) {
            return intValue
        }
        // If that fails, try to decode as String and convert
        if let stringValue = try? decodeIfPresent(String.self, forKey: key) {
            // Return nil for empty strings
            if stringValue.isEmpty {
                return nil
            }
            // Try to convert string to int
            return Int(stringValue)
        }
        return nil
    }
}

struct Property: Identifiable, Codable {
    let id: String
    let title: String
    let description: String
    let address: String
    let city: String
    let state: String
    let zipCode: String
    let county: String?
    let country: String
    let status: PropertyStatus
    let createdAt: Date
    let updatedAt: Date
    let owner: String
    let ownerDetails: User
    let mediaCount: Int
    let descriptives: PropertyDescriptives?
    let serviceRequests: [ServiceRequest]?
    let geocodeAddress: GeocodeAddress?
    let geocodeAddressSource: String?
    let permitRetrievalStatus: String?
    let permitLastRetrievedAt: Date?
    let permitRetrievalError: String?
    let permitNextRetrievalAt: Date?
    let permitRetrievalWorkflowId: String?
    private let _featuredImage: String?
    private let _media: [Media]?
    
    // Computed properties that rewrite media URLs
    var featuredImage: String? {
        guard let image = _featuredImage else { return nil }
        return NetworkManager.shared.rewriteMediaURL(image)
    }
    
    var media: [Media]? {
        return _media?.map { mediaItem in
            var updatedItem = mediaItem
            if !mediaItem.fileUrl.isEmpty {
                updatedItem.fileUrl = NetworkManager.shared.rewriteMediaURL(mediaItem.fileUrl)
            }
            return updatedItem
        }
    }
    
    // UI properties
    var image: UIImage?
    
    // Custom initializer to handle the renamed properties
    init(
        id: String,
        title: String,
        description: String,
        address: String,
        city: String,
        state: String,
        zipCode: String,
        county: String?,
        country: String,
        status: PropertyStatus,
        createdAt: Date,
        updatedAt: Date,
        owner: String,
        ownerDetails: User,
        mediaCount: Int,
        descriptives: PropertyDescriptives?,
        serviceRequests: [ServiceRequest]?,
        geocodeAddress: GeocodeAddress?,
        geocodeAddressSource: String?,
        permitRetrievalStatus: String?,
        permitLastRetrievedAt: Date?,
        permitRetrievalError: String?,
        permitNextRetrievalAt: Date?,
        permitRetrievalWorkflowId: String?,
        featuredImage: String?,
        media: [Media]?
    ) {
        self.id = id
        self.title = title
        self.description = description
        self.address = address
        self.city = city
        self.state = state
        self.zipCode = zipCode
        self.county = county
        self.country = country
        self.status = status
        self.createdAt = createdAt
        self.updatedAt = updatedAt
        self.owner = owner
        self.ownerDetails = ownerDetails
        self.mediaCount = mediaCount
        self.descriptives = descriptives
        self.serviceRequests = serviceRequests
        self.geocodeAddress = geocodeAddress
        self.geocodeAddressSource = geocodeAddressSource
        self.permitRetrievalStatus = permitRetrievalStatus
        self.permitLastRetrievedAt = permitLastRetrievedAt
        self.permitRetrievalError = permitRetrievalError
        self.permitNextRetrievalAt = permitNextRetrievalAt
        self.permitRetrievalWorkflowId = permitRetrievalWorkflowId
        self._featuredImage = featuredImage
        self._media = media
    }
    
    enum CodingKeys: String, CodingKey {
        case id
        case title
        case description
        case address
        case city
        case state
        case zipCode
        case county
        case country
        case status
        case createdAt
        case updatedAt
        case owner
        case ownerDetails
        case mediaCount
        case descriptives
        case serviceRequests
        case geocodeAddress
        case geocodeAddressSource
        case permitRetrievalStatus
        case permitLastRetrievedAt
        case permitRetrievalError
        case permitNextRetrievalAt
        case permitRetrievalWorkflowId
        case _featuredImage = "featuredImage"
        case _media = "media"
    }
}

enum PropertyStatus: String, Codable {
    case ACTIVE
    case INACTIVE
    case PENDING
    case COUNTY_PROCESSING
    case UNKNOWN
    
    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        let rawValue = try container.decode(String.self)
        self = PropertyStatus(rawValue: rawValue) ?? .UNKNOWN
    }
}

struct PropertyDescriptives: Codable {
    // Basic Property Info
    let propertyType: String?
    let yearBuilt: Int?
    let squareFootage: Int?
    let lotSize: String?
    let stories: Int?
    let bedrooms: Int?
    let bathrooms: String?
    let unitNumber: String?
    
    // Access & Security
    let gatedCommunity: Bool?
    let accessCode: String?
    let accessInstructions: String?
    let parkingType: String?
    let parkingSpaces: Int?
    
    // Structure & Features
    let basement: Bool?
    let basementType: String?
    let garage: Bool?
    let garageType: String?
    let garageSpaces: Int?
    let attic: Bool?
    let atticAccess: String?
    let crawlSpace: Bool?
    
    // HVAC & Climate Control
    let heatingSystem: String?
    let heatingFuel: String?
    let coolingSystem: String?
    let airConditioning: Bool?
    let hvacAge: Int?
    let hvacBrand: String?
    let hvacModel: String?
    let thermostatType: String?
    
    // Utilities & Systems
    let utilities: PropertyUtilities?
    let waterSource: String?
    let sewerSystem: String?
    let electricalPanel: String?
    let electricalAmps: Int?
    let gasService: Bool?
    let waterHeater: PropertyWaterHeater?
    
    // Plumbing
    let plumbingType: String?
    let waterShutoffLocation: String?
    let mainDrainCleanout: String?
    
    // Electrical
    let electricalPanelLocation: String?
    let wiringType: String?
    
    // Roofing & Exterior
    let roofType: String?
    let roofAge: String?
    let exteriorMaterial: String?
    let foundationType: String?
    
    // Appliances
    let appliances: PropertyAppliances?
    
    // Smart Home & Internet
    let internetService: Bool?
    let internetProvider: String?
    let smartHomeDevices: [String]?
    
    // Landscaping & Exterior
    let sprinklerSystem: Bool?
    let pool: Bool?
    let poolType: String?
    let fence: Bool?
    let fenceType: String?
    let deck: Bool?
    let deckMaterial: String?
    let patio: Bool?
    let patioMaterial: String?
    
    // Special Considerations
    let petFriendly: Bool?
    let smokingAllowed: Bool?
    let wheelchairAccessible: Bool?
    let fireplace: Bool?
    let fireplaceType: String?
    
    // Maintenance & Notes
    let maintenanceNotes: String?
    let specialInstructions: String?
    
    // Legacy field
    let createdFrom: String?
    
    // Custom decoder to handle flexible Int/String types from backend
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        
        // Decode strings
        propertyType = try container.decodeIfPresent(String.self, forKey: .propertyType)
        lotSize = try container.decodeIfPresent(String.self, forKey: .lotSize)
        bathrooms = try container.decodeIfPresent(String.self, forKey: .bathrooms)
        unitNumber = try container.decodeIfPresent(String.self, forKey: .unitNumber)
        accessCode = try container.decodeIfPresent(String.self, forKey: .accessCode)
        accessInstructions = try container.decodeIfPresent(String.self, forKey: .accessInstructions)
        parkingType = try container.decodeIfPresent(String.self, forKey: .parkingType)
        basementType = try container.decodeIfPresent(String.self, forKey: .basementType)
        garageType = try container.decodeIfPresent(String.self, forKey: .garageType)
        atticAccess = try container.decodeIfPresent(String.self, forKey: .atticAccess)
        heatingSystem = try container.decodeIfPresent(String.self, forKey: .heatingSystem)
        heatingFuel = try container.decodeIfPresent(String.self, forKey: .heatingFuel)
        coolingSystem = try container.decodeIfPresent(String.self, forKey: .coolingSystem)
        hvacBrand = try container.decodeIfPresent(String.self, forKey: .hvacBrand)
        hvacModel = try container.decodeIfPresent(String.self, forKey: .hvacModel)
        thermostatType = try container.decodeIfPresent(String.self, forKey: .thermostatType)
        waterSource = try container.decodeIfPresent(String.self, forKey: .waterSource)
        sewerSystem = try container.decodeIfPresent(String.self, forKey: .sewerSystem)
        electricalPanel = try container.decodeIfPresent(String.self, forKey: .electricalPanel)
        plumbingType = try container.decodeIfPresent(String.self, forKey: .plumbingType)
        waterShutoffLocation = try container.decodeIfPresent(String.self, forKey: .waterShutoffLocation)
        mainDrainCleanout = try container.decodeIfPresent(String.self, forKey: .mainDrainCleanout)
        electricalPanelLocation = try container.decodeIfPresent(String.self, forKey: .electricalPanelLocation)
        wiringType = try container.decodeIfPresent(String.self, forKey: .wiringType)
        roofType = try container.decodeIfPresent(String.self, forKey: .roofType)
        roofAge = try container.decodeIfPresent(String.self, forKey: .roofAge)
        exteriorMaterial = try container.decodeIfPresent(String.self, forKey: .exteriorMaterial)
        foundationType = try container.decodeIfPresent(String.self, forKey: .foundationType)
        internetProvider = try container.decodeIfPresent(String.self, forKey: .internetProvider)
        poolType = try container.decodeIfPresent(String.self, forKey: .poolType)
        fenceType = try container.decodeIfPresent(String.self, forKey: .fenceType)
        deckMaterial = try container.decodeIfPresent(String.self, forKey: .deckMaterial)
        patioMaterial = try container.decodeIfPresent(String.self, forKey: .patioMaterial)
        fireplaceType = try container.decodeIfPresent(String.self, forKey: .fireplaceType)
        maintenanceNotes = try container.decodeIfPresent(String.self, forKey: .maintenanceNotes)
        specialInstructions = try container.decodeIfPresent(String.self, forKey: .specialInstructions)
        createdFrom = try container.decodeIfPresent(String.self, forKey: .createdFrom)
        
        // Decode flexible Int fields (can be Int or String in JSON)
        yearBuilt = try container.decodeFlexibleInt(forKey: .yearBuilt)
        squareFootage = try container.decodeFlexibleInt(forKey: .squareFootage)
        stories = try container.decodeFlexibleInt(forKey: .stories)
        bedrooms = try container.decodeFlexibleInt(forKey: .bedrooms)
        parkingSpaces = try container.decodeFlexibleInt(forKey: .parkingSpaces)
        garageSpaces = try container.decodeFlexibleInt(forKey: .garageSpaces)
        hvacAge = try container.decodeFlexibleInt(forKey: .hvacAge)
        electricalAmps = try container.decodeFlexibleInt(forKey: .electricalAmps)
        
        // Decode booleans
        gatedCommunity = try container.decodeIfPresent(Bool.self, forKey: .gatedCommunity)
        basement = try container.decodeIfPresent(Bool.self, forKey: .basement)
        garage = try container.decodeIfPresent(Bool.self, forKey: .garage)
        attic = try container.decodeIfPresent(Bool.self, forKey: .attic)
        crawlSpace = try container.decodeIfPresent(Bool.self, forKey: .crawlSpace)
        airConditioning = try container.decodeIfPresent(Bool.self, forKey: .airConditioning)
        gasService = try container.decodeIfPresent(Bool.self, forKey: .gasService)
        internetService = try container.decodeIfPresent(Bool.self, forKey: .internetService)
        sprinklerSystem = try container.decodeIfPresent(Bool.self, forKey: .sprinklerSystem)
        pool = try container.decodeIfPresent(Bool.self, forKey: .pool)
        fence = try container.decodeIfPresent(Bool.self, forKey: .fence)
        deck = try container.decodeIfPresent(Bool.self, forKey: .deck)
        patio = try container.decodeIfPresent(Bool.self, forKey: .patio)
        petFriendly = try container.decodeIfPresent(Bool.self, forKey: .petFriendly)
        smokingAllowed = try container.decodeIfPresent(Bool.self, forKey: .smokingAllowed)
        wheelchairAccessible = try container.decodeIfPresent(Bool.self, forKey: .wheelchairAccessible)
        fireplace = try container.decodeIfPresent(Bool.self, forKey: .fireplace)
        
        // Decode nested objects
        utilities = try container.decodeIfPresent(PropertyUtilities.self, forKey: .utilities)
        waterHeater = try container.decodeIfPresent(PropertyWaterHeater.self, forKey: .waterHeater)
        appliances = try container.decodeIfPresent(PropertyAppliances.self, forKey: .appliances)
        smartHomeDevices = try container.decodeIfPresent([String].self, forKey: .smartHomeDevices)
    }
    
    // Convenience initializer for common fields
    init(
        propertyType: String? = nil,
        yearBuilt: Int? = nil,
        squareFootage: Int? = nil,
        bedrooms: Int? = nil,
        bathrooms: String? = nil,
        unitNumber: String? = nil,
        garage: Bool? = nil,
        basement: Bool? = nil,
        gatedCommunity: Bool? = nil,
        heatingSystem: String? = nil,
        airConditioning: Bool? = nil,
        utilities: PropertyUtilities? = nil,
        createdFrom: String? = nil
    ) {
        self.propertyType = propertyType
        self.yearBuilt = yearBuilt
        self.squareFootage = squareFootage
        self.lotSize = nil
        self.stories = nil
        self.bedrooms = bedrooms
        self.bathrooms = bathrooms
        self.unitNumber = unitNumber
        self.gatedCommunity = gatedCommunity
        self.accessCode = nil
        self.accessInstructions = nil
        self.parkingType = nil
        self.parkingSpaces = nil
        self.basement = basement
        self.basementType = nil
        self.garage = garage
        self.garageType = nil
        self.garageSpaces = nil
        self.attic = nil
        self.atticAccess = nil
        self.crawlSpace = nil
        self.heatingSystem = heatingSystem
        self.heatingFuel = nil
        self.coolingSystem = nil
        self.airConditioning = airConditioning
        self.hvacAge = nil
        self.hvacBrand = nil
        self.hvacModel = nil
        self.thermostatType = nil
        self.utilities = utilities
        self.waterSource = nil
        self.sewerSystem = nil
        self.electricalPanel = nil
        self.electricalAmps = nil
        self.gasService = nil
        self.waterHeater = nil
        self.plumbingType = nil
        self.waterShutoffLocation = nil
        self.mainDrainCleanout = nil
        self.electricalPanelLocation = nil
        self.wiringType = nil
        self.roofType = nil
        self.roofAge = nil
        self.exteriorMaterial = nil
        self.foundationType = nil
        self.appliances = nil
        self.internetService = nil
        self.internetProvider = nil
        self.smartHomeDevices = nil
        self.sprinklerSystem = nil
        self.pool = nil
        self.poolType = nil
        self.fence = nil
        self.fenceType = nil
        self.deck = nil
        self.deckMaterial = nil
        self.patio = nil
        self.patioMaterial = nil
        self.petFriendly = nil
        self.smokingAllowed = nil
        self.wheelchairAccessible = nil
        self.fireplace = nil
        self.fireplaceType = nil
        self.maintenanceNotes = nil
        self.specialInstructions = nil
        self.createdFrom = createdFrom
    }
    
    enum CodingKeys: String, CodingKey {
        case propertyType
        case yearBuilt
        case squareFootage
        case lotSize
        case stories
        case bedrooms
        case bathrooms
        case unitNumber
        case gatedCommunity
        case accessCode
        case accessInstructions
        case parkingType
        case parkingSpaces
        case basement
        case basementType
        case garage
        case garageType
        case garageSpaces
        case attic
        case atticAccess
        case crawlSpace
        case heatingSystem
        case heatingFuel
        case coolingSystem
        case airConditioning
        case hvacAge
        case hvacBrand
        case hvacModel
        case thermostatType
        case utilities
        case waterSource
        case sewerSystem
        case electricalPanel
        case electricalAmps
        case gasService
        case waterHeater
        case plumbingType
        case waterShutoffLocation
        case mainDrainCleanout
        case electricalPanelLocation
        case wiringType
        case roofType
        case roofAge
        case exteriorMaterial
        case foundationType
        case appliances
        case internetService
        case internetProvider
        case smartHomeDevices
        case sprinklerSystem
        case pool
        case poolType
        case fence
        case fenceType
        case deck
        case deckMaterial
        case patio
        case patioMaterial
        case petFriendly
        case smokingAllowed
        case wheelchairAccessible
        case fireplace
        case fireplaceType
        case maintenanceNotes
        case specialInstructions
        case createdFrom = "created_from"
    }
}

struct PropertyUtilities: Codable {
    let gas: String?
    let sewer: String?
    let water: String?
    let electricity: String?
    let internetCable: String?
    
    enum CodingKeys: String, CodingKey {
        case gas
        case sewer
        case water
        case electricity
        case internetCable
    }
}

struct PropertyWaterHeater: Codable {
    let type: String?
    let fuel: String?
    let capacity: String?
    let age: Int?
    let brand: String?
    let model: String?
}

struct PropertyAppliances: Codable {
    let refrigerator: Bool?
    let stove: Bool?
    let dishwasher: Bool?
    let washer: Bool?
    let dryer: Bool?
    let microwave: Bool?
    let oven: Bool?
    let range: Bool?
    let garbageDisposal: Bool?
}

struct GeocodeAddress: Codable {
    let results: [GeocodeResult]?
    let summary: GeocodeSummary?
    
    struct GeocodeResult: Codable {
        let id: String?
        let type: String?
        let score: Double?
        let address: GeocodeResultAddress?
        let position: GeocodePosition?
    }
    
    struct GeocodeResultAddress: Codable {
        let country: String?
        let localName: String?
        let postalCode: String?
        let streetName: String?
        let countryCode: String?
        let municipality: String?
        let streetNumber: String?
        let freeformAddress: String?
    }
    
    struct GeocodePosition: Codable {
        let lat: Double?
        let lon: Double?
    }
    
    struct GeocodeSummary: Codable {
        let query: String?
        let queryTime: Int?
    }
}